File: native-plugins/admob/resources/ios/Sources/AppOpenRegistry.swift
import Foundation
import GoogleMobileAds

/// Slot-keyed registry for loaded `AppOpenAd` instances, their delegates,
/// and their load timestamps for staleness checks.
///
/// Google strongly recommends discarding app-open ads older than 4 hours;
/// `isFresh(slot:)` enforces this.
///
/// The delegate map exists for the same retention reason as the
/// InterstitialRegistry - `FullScreenContentDelegate` is weakly held by the
/// SDK, so the registry must strongly retain it until dismissal.
final class AppOpenRegistry {
    static let shared = AppOpenRegistry()

    private var ads: [String: AppOpenAd] = [:]
    private var delegates: [String: AppOpenDelegate] = [:]
    private var loadTimes: [String: Date] = [:]

    private static let staleThreshold: TimeInterval = 4 * 60 * 60   // 4 hours

    private init() {}

    func put(slot: String, ad: AppOpenAd, delegate: AppOpenDelegate) {
        ads[slot] = ad
        delegates[slot] = delegate
        loadTimes[slot] = Date()
    }

    func get(slot: String) -> AppOpenAd? {
        return ads[slot]
    }

    func remove(slot: String) {
        ads.removeValue(forKey: slot)
        delegates.removeValue(forKey: slot)
        loadTimes.removeValue(forKey: slot)
    }

    func isFresh(slot: String) -> Bool {
        guard let t = loadTimes[slot] else { return false }
        return Date().timeIntervalSince(t) < AppOpenRegistry.staleThreshold
    }

    func ageMs(slot: String) -> Int {
        guard let t = loadTimes[slot] else { return -1 }
        return Int(Date().timeIntervalSince(t) * 1000)
    }

    func allSlots() -> [String] {
        return Array(ads.keys)
    }

    func clear() {
        ads.removeAll()
        delegates.removeAll()
        loadTimes.removeAll()
    }
}

File: native-plugins/admob/resources/android/src/AppOpenRegistry.kt
package com.blessedzulu.nativephp.admob

import com.google.android.gms.ads.appopen.AppOpenAd

/**
 * Slot-keyed registry for loaded AppOpenAd instances plus their load
 * timestamps for staleness checks.
 *
 * Google strongly recommends discarding app-open ads older than 4 hours -
 * stale ones drift attribution and serve poorly. `isFresh(slot)` enforces
 * this threshold.
 *
 * One-shot: the slot is cleared in dismissal callbacks. The AppOpenLifecycle
 * observer also clears stale slots so the next lifecycle resume finds them
 * empty and either does nothing or triggers a fresh load.
 *
 * Thread-safety: all public methods are synchronized.
 */
object AppOpenRegistry {
    private val ads = mutableMapOf<String, AppOpenAd>()
    private val loadTimes = mutableMapOf<String, Long>()

    private const val STALE_THRESHOLD_MS = 4L * 60L * 60L * 1000L

    @Synchronized
    fun put(slot: String, ad: AppOpenAd) {
        ads[slot] = ad
        loadTimes[slot] = System.currentTimeMillis()
    }

    @Synchronized
    fun get(slot: String): AppOpenAd? = ads[slot]

    @Synchronized
    fun remove(slot: String): AppOpenAd? {
        loadTimes.remove(slot)
        return ads.remove(slot)
    }

    @Synchronized
    fun isFresh(slot: String): Boolean {
        val t = loadTimes[slot] ?: return false
        return (System.currentTimeMillis() - t) < STALE_THRESHOLD_MS
    }

    @Synchronized
    fun ageMs(slot: String): Long? = loadTimes[slot]?.let { System.currentTimeMillis() - it }

    @Synchronized
    fun allSlots(): List<String> = ads.keys.toList()

    @Synchronized
    fun clear() {
        ads.clear()
        loadTimes.clear()
    }
}
